pennylane/ops/op_math/symbolicop.py

```diff
@@ -258,27 +258,30 @@ def matrix(self, wire_order=None):
             base_matrix = self.base.matrix()
 
         scalar_interface = qml.math.get_interface(self.scalar)
+        scalar = self.scalar
         if scalar_interface == "torch":
             # otherwise get `RuntimeError: Can't call numpy() on Tensor that requires grad.`
             base_matrix = qml.math.convert_like(base_matrix, self.scalar)
-        elif scalar_interface == "tensorflow" and not qml.math.iscomplex(self.scalar):
-            # cast scalar to complex to avoid an error
-            self.scalar = qml.math.cast_like(self.scalar, base_matrix)
+        elif scalar_interface == "tensorflow":
```

Comment:
not sure if this change will have a significant performance impact, but we used to only cast once and update `self.scalar` in place. with this, every call to `op.matrix()` will require two calls to `qml.math.cast`.

Before, when `self.scalar` was changed in place:
```pycon
>>> op=qml.s_prod(tf.Variable(1.1), qml.PauliY(0))
>>> op.scalar
<tf.Variable 'Variable:0' shape=() dtype=float32, numpy=1.1>
>>> op.matrix()
<tf.Tensor: shape=(2, 2), dtype=complex128, numpy=
array([[0.+0.j        , 0.-1.10000002j],
       [0.+1.10000002j, 0.+0.j        ]])>
>>> op.scalar
<tf.Tensor: shape=(), dtype=complex128, numpy=(1.100000023841858+0j)>
```

Reply:
I timed 50k subsequent calls to `op.matrix()` with the above op, and it took 200+/-30 $\mu s$ on master, but 90+/10 $\mu s$ on this branch so I stand corrected! not sure how that's happening...